File: coin change2 dp.java

// time complexiy of this problem is O(m*n) (time is m*n proportional since we are filling the matrix) and space complexity O(m*n) 
// this program works well
// here we the amount is less than the coin denomination than we take it from above ie already calculated else we are adding the above calculated value and going back to the column by the number of coin denomination and then adding that value to that previous and the  returning the last value of the matrix.
class Solution {
    public int change(int amount, int[] coins) {
        int  [][] dp= new int [coins.length+1][amount+1];
        dp[0][0]=1;
        int r=dp.length; // no of rows
        int c=dp[0].length; // no of columns
        for(int i=1;i<r;i++)
        {
            dp[i][0]=1;
        
        for(int j=1;j<c;j++)
        {
            if(j<coins[i-1])
            {
                dp[i][j]=dp[i-1][j];
            }
            else
                dp[i][j]=dp[i-1][j]+ dp[i][j-coins[i-1]];
        }
        }
       return dp[r-1][c-1];
    }

}

File: coin_change_2_dp.cpp
/*
 * @Author: Yuvaraj Nagi
 * @Date: 2022-11-16 18:55:59
 * @Last Modified by: Yuvaraj Nagi
 * @Last Modified time: 2022-11-16 19:35:50
 * Time complexity : O(m x n)
 * Space Complexity : O(m x n)
 *
 */

#include <vector>

using namespace std;
class Solution
{
public:
    int change(int amount, vector<int> &coins)
    {
        // coin change 2 : using dynamic programming to solve our problem:
        // first declare a matrix of rows of size amount and columns of size coins

        int r = coins.size() + 1;
        int c = amount + 1;

        int matrix[r][c];

        // initializing everything after first element in row 1 to 0
        for (int i = 0; i < c; i++)
        {
            matrix[0][i] = 0;
        }

        // initializing everything in row 1 to 1
        for (int j = 0; j < r; j++)
        {
            matrix[j][0] = 1;
        }

        for (int i = 1; i < r; i++)
        {
            for (int j = 1; j < c; j++)
            {
                if (j < coins[i - 1])
                {
                    // column index is lesser than coins array element
                    matrix[i][j] = matrix[i - 1][j];
                }

                else
                {
                    matrix[i][j] = matrix[i - 1][j] + matrix[i][j - coins[i - 1]];
                }
            }
        }

        return matrix[r - 1][c - 1];
    }
};